seq2science/util.py

```diff
@@ -407,3 +408,132 @@ def shorten(string, max_length, methods="right"):
+
+
+def parse_samples(config):
+    from pandas_schema import Column, Schema
+    from pandas_schema.validation import MatchesPatternValidation, IsDistinctValidation
+    from snakemake.utils import validate
+
+    # read the samples.tsv file as all text, drop comment lines
+    samples = pd.read_csv(config["samples"], sep='\t', dtype='str', comment='#')
+    samples.columns = samples.columns.str.strip()
+
+    assert all([col[0:7] not in ["Unnamed", ''] for col in samples]), \
+        (f"\nEncountered unnamed column in {config['samples']}.\n" +
+         f"Column names: {str(', '.join(samples.columns))}.\n")
+
+    # use pandasschema for checking if samples file is filed out correctly
+    allowed_pattern = r'[A-Za-z0-9_.\-%]+'
+    distinct_columns = ["sample"]
+    if "descriptive_name" in samples.columns:
+        distinct_columns.append("descriptive_name")
+
+    distinct_schema = Schema(
+        [Column(col, [MatchesPatternValidation(allowed_pattern),
+                      IsDistinctValidation()] if col in distinct_columns else [
+            MatchesPatternValidation(allowed_pattern)], allow_empty=True) for col in
+         samples.columns])
+
+    errors = distinct_schema.validate(samples)
+
+    if len(errors):
+        logger.error("\nThere are some issues with parsing the samples file:")
+        for error in errors:
+            logger.error(error)
+        logger.error("")  # empty line
+        raise TerminatedException
+
+    # for each column, if found in samples.tsv:
+    # 1) if it is incomplete, fill the blanks with replicate/sample names
+    # (sample names if replicates are not found/applicable)
+    # 2) drop column if it is identical to the replicate/sample column, or if not needed
+    if 'replicate' in samples:
+        samples['replicate'] = samples['replicate'].mask(pd.isnull, samples['sample'])
+        if samples['replicate'].tolist() == samples['sample'].tolist() or config.get('technical_replicates') == 'keep':
+            samples = samples.drop(columns=['replicate'])
+    if 'condition' in samples:
+        samples['condition'] = samples['condition'].mask(pd.isnull, samples['replicate']) if 'replicate' in samples else \
+            samples['condition'].mask(pd.isnull, samples['sample'])
+        if samples['condition'].tolist() == samples['sample'].tolist() or config.get('biological_replicates') == 'keep':
+            samples = samples.drop(columns=['condition'])
+    if 'descriptive_name' in samples:
+        samples['descriptive_name'] = samples['descriptive_name'].mask(pd.isnull, samples['replicate']) if \
+            'replicate' in samples else samples['descriptive_name'].mask(pd.isnull, samples['sample'])
+        if ('replicate' in samples and samples['descriptive_name'].to_list() == samples['replicate'].to_list()) or \
+                samples['descriptive_name'].to_list() == samples['sample'].to_list():
+            samples = samples.drop(columns=['descriptive_name'])
```

Comment:
these two were updated.

Reply:
good catch, The merge just auto-merged this stuff :eyes:

Reply:
why does this not make some tests fail?! :warning:

Reply:
not used... :facepalm: